Issue #45: Add Camelot Nitro Pool Interface to SDK

jahabeebs opened the pull request:
Closes #45 
Closes #56 
Closes #53 
Closes #54 
Closes #55 

## Description
- I added the function fetchNitroPoolODGwstETH (...maybe I should snake case this) which fetches pool data from an ODG-WSTETH nitro pool and returns a lot of relevant stats
- I added several new contracts to our contract API and GEB object (note that I added the ABIs into our local folder)--this will allow us to generate typechained definitions for these contracts and interact with them in the app easily. These contracts include:
1) camelotNitroPool, the camelot nitro pool ABI which I've taken from Lyra Finance (https://github.com/lyra-finance/interface/commit/e60e7bcf9d43ebafa8ee1b9d6ba8820e7572d00e#diff-01893f21d6faed4a7146ec796292868d585f39615ccd5787961a9e5a4e5f65da), I'm assuming it's up to date for now until we deploy our own
2) camelotODGPool, this is of type CamelotPool which is the ABI of our *deployed ODG pool*. We need this so that we can access market price info for ODG. However, there's a problem with this contract which I'll note in a section below
3) mutlicall, this is a contract that lets us create batch calls on any network--we already had the addresses for this multicall contract in our addresses.ts file, however it wasn't available on the GEB object or our contract APIs so I used the official multicall site to get the ABI and this will let us do batch calls in the app

## Problems/questions

We're missing three key pieces of info for testing this code
1) No deployed Arbitrum Sepolia Camelot pools for ODG-wstETH or ODG-wETH. I've added a TODO in addresses.ts for adding these two contracts 
2) Need chainlinkRelayer for wstETH—in fetchPoolData we're using chainlinkRelayer to get the wETH market price. However, we do not have a chainlinkRelayer for Arbitrum Sepolia, just Arbitrum Goerli, which which is probably why failures occur with fetching pool data in our app
3) The deployed CamelotPool contract we're using for ODG has a lot of functions but not totalSupply. We need a Camelot pool contract with totalSupply as part of our multicall logic. I don't know why totalSupply is missing from our contract -- maybe we're using a new version? The CamelotPool ABI that Lyra is using just has totalSupply (https://github.com/lyra-finance/interface/blob/e60e7bcf9d43ebafa8ee1b9d6ba8820e7572d00e/app/src/contracts/abis/CamelotPool.json). Is there an alternative way to calculate this? I've added a TODO in fetchNitroPool.ts to remind us that the ABI needs to be addressed

Also, we'll later need to create a function for fetching from the RETH pool, but we should get this one working first

jahabeebs committed:
draft adding camelot nitro pool interface

jahabeebs committed:
rename function

jahabeebs committed:
fix types

pi0neerpat commented on src/contracts/addreses.ts:
```diff
@@ -78,6 +80,9 @@ const addresses: Record<GebDeployment, ContractList> = {
+        CAMELOT_NITRO_POOL: '',
+        CAMELOT_ODG_POOL: ''
     },
     'arbitrum-goerli': {
```
Are you on the latest branch? I thought I removed `arbitrum-goerli`

jahabeebs replied:
yes, goerli is still in dev

pi0neerpat commented on src/contracts/addreses.ts:
```diff
@@ -37,6 +37,8 @@ export type ContractKey =
     | 'JOB_ACCOUNTING'
     | 'JOB_LIQUIDATION'
     | 'JOB_ORACLES'
+    | 'CAMELOT_NITRO_POOL'
```
Will need to differentiate Nitro pools by collateral types: `RETH`, `WSTETH`

jahabeebs replied:
hmm good thought, I'll add that in

pi0neerpat commented on src/camelot/utils.ts:
Maybe move to `src/`, since this may be used for other features

jahabeebs committed:
move utils, add separate pool addresses

jahabeebs committed:
update pool calculations and nitro abi

jahabeebs commented:
Pushed several updates:
- Gets rid of totalSupply logic
- New (rough) APY calculation: ```(rewardsPerSecond * SECONDS_IN_YEAR * odgMarketPriceFloat) / tvl```
- Updates Nitro Pool ABI with more functions, now we get this pool-level data:
```
   settings: {
        startTime: BigNumber
        endTime: BigNumber
        harvestStartTime: BigNumber
        depositEndTime: BigNumber
        lockDurationReq: BigNumber
        lockEndReq: BigNumber
        depositAmountReq: BigNumber
        whitelist: boolean
        description: string
    }
```
- Returns the entire userInfo object to get important deposit info:
```
    userInfo: {
        totalDepositAmount: BigNumber,
        rewardDebtToken1: BigNumber,
        rewardDebtToken2: BigNumber,
        pendingRewardsToken1: BigNumber,
        pendingRewardsToken2: BigNumber
    } | null,
```

pi0neerpat committed:
Merge branch 'dev' into feature/45-add-camelot-nitro-pool-interface

pi0neerpat committed:
MulticallRequest export fix

jahabeebs committed:
update sdk logic

jahabeebs commented:
Pushed several updates:
- Removed camelotODGPool (we will have wsteth and reth pools, this is not needed)
- Renamed MulticallRequest to CamelotMulticallRequest
- balanceOf logic changed in fetchNitroPool logic to pass in correct camelotwstETHNitroPool address to the test ODG and wstETH contracts
- Addresses updated to include the correct test contracts (like the wstETH pool, test wstETH, and test ODG addresses)

pi0neerpat committed:
Merge branch 'dev' into feature/45-add-camelot-nitro-pool-interface

pi0neerpat committed:
cleanup addresses

pi0neerpat committed:
temporarily mock market prices

pi0neerpat committed:
temporarily remove lpTokenBalance and userInfo

pi0neerpat reviewed (changes requested):
I pushed some changes to get this to display details about the pool in the app. See https://github.com/open-dollar/od-app/pull/252 

TODO

- Add back data for `lpTokenBalance` and `userInfo`, which I temporarily removed while debugging
- The naming convention for the SDK is to make token symbols always uppercase. Eg fetchNitroPool wstETH_market_price should be changed to WSTETH_market_price

Other comments that I already addressed:

- Use 0x0 addresses instead of a real address from the wrong network. This prevents confusion in the future.

pi0neerpat committed:
add mainnet collateral addresses

pi0neerpat committed:
lint

jahabeebs committed:
refactor token names, fix user info and lp token fetch

jahabeebs commented:
@pi0neerpat 
- Added back data for lpTokenBalance and userInfo
- Renamed wstETH to WSTETH throughout my code so that it adheres to standards

Here's what the stringified data looks like now in the app

<img width="1301" alt="Screenshot 2024-02-05 at 2 25 22 PM" src="https://github.com/open-dollar/od-sdk/assets/47253537/bf8783b1-3814-45d1-9452-ae10775bcc94">

pi0neerpat reviewed (commented):
What data is left to gather based on the designs?

![image](https://github.com/open-dollar/od-sdk/assets/35622595/ec012586-fc68-43bc-99ac-408c28eb1c5c)

jahabeebs commented:
> What data is left to gather based on the designs?
> 
> ![image](https://private-user-images.githubusercontent.com/35622595/302425331-ec012586-fc68-43bc-99ac-408c28eb1c5c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcxNjczOTIsIm5iZiI6MTcwNzE2NzA5MiwicGF0aCI6Ii8zNTYyMjU5NS8zMDI0MjUzMzEtZWMwMTI1ODYtZmM2OC00M2JjLTk5YWMtNDA4YzI4ZWIxYzVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA1VDIxMDQ1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZiNzk0NDMyMTQxOGFlYWQ2MzIyZGU2YTcwNWUzMTYyNzBlODEyMmYwZmZjMzM4MmQ1NmY2ODhhNTJiNjE1ZmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.CIIOjOij-brx1MBp6GwlZjQJ4zDbc1LqqFe6JBGw150)

Nothing, this has 100% of the data needed

We just need to replace the price data with real values

jahabeebs committed:
refactored fetchNitroPool working with wstETH and RETH

jahabeebs commented:
I pushed changes to this PR that refactor the nitro pool method to fetch data for our RETH or WSTETH test pools. so in the app for the RETH-ODG pool you would call it like 

```
const data = await fetchNitroPool(geb, 'RETH','0xaddresshere')
```

A strange thing that happened was when testing today I was getting errors pulling from my wstETH pool until I re-imported the token to camelot's UI, might need to do some more testing to see if pulling from our pools long-term relies on Camelot allow listing a token

This PR now closes #53 too

## Screenshots

https://github.com/open-dollar/od-sdk/assets/47253537/c2c5ac68-2389-497e-aa63-4c7e779eec18

jahabeebs committed:
add chainlink relayer to reth, wsteth, odg, add available collateral function, add pending rewards

jahabeebs committed:
available deposit types

jahabeebs commented:
Pushed an update that includes the following:

1) no hardcoded prices--RETH, ODG, and WSTETH prices are being fetched from a USD denominated oracle contract that I added to ```addresses.ts```. The only issue here is that the validity of RETH and WSTETH is coming back as false so I had to temporarily disable the validity check of these prices. This should be ok for now since it's read-only data
2) ```availableCollateralsForDeposit``` function returns an array of available collaterals for deposit
3) added pendingRewards object to response which includes pending1 and pending2 values...there are two values within the object because through the Camelot UI we can add incentives twice. in the example pools I've only added incentives once so only pending1 will be populated
4) Added type AvailableDepositTypes to prevent type casting in the app

This now closes #54 and #55--we didn't need to use the coingecko fetch since now we have USD denominated oracle contracts

jahabeebs commented:
Closing as won't do -- part of TVL project
